R/defStormsDataset.R

```diff
@@ -241,23 +248,350 @@ checkInputsdefStormsDataset <- function(filename, fields, basin, seasons, unitCo
+    } else {
+      data$poci <- array(ncdf4::ncvar_get(dataBase, fields["poci"])[, ind], dim = c(row, len))
+    }
+    data$poci <- data$poci[, o]
+  }
+  
+  
+  ncdf4::nc_close(dataBase)
+  
+  return(data)
+}
+
+
+
+
+#' Load database when filename is a NetCDF
+#' (CF defStormsDataset for additional informations about parameters)
+#'
+#' @param filename character
+#' @param fields named character vecor
+#' @param basin character
+#' @param seasons numeric vector
+#' @param unitConversion named character vector
+#' @param verbose numeric
+#'
+#' @return list of arrays
+getDataFromCsvFile <- function(filename, fields, basin, seasons, unitConversion, verbose){
+  
+  if (verbose) {
+    cat("=== Loading data  ===\nOpen database... ")
+  }
+  
+  dataBase <- utils::read.csv(filename)
+  
+  if (verbose) {
+    cat(filename, "opened\nCollecting data ...\n")
+  }
+  
+  # Extract columns
+  filter <- which(colnames(dataBase) %in% fields)
+  dataBaseFiltered <- dataBase[, filter]
+  
+  
+  # Remove sub header
+  dataBaseFiltered <- dataBase[2:dim(dataBaseFiltered)[1], filter]
+ 
+  
+  # filter by season
+  filter <- which(as.numeric(dataBaseFiltered[,fields["seasons"]]) >= seasons[1] & as.numeric(dataBaseFiltered[,fields["seasons"]]) <= seasons[2])
+  dataBaseFiltered <- dataBaseFiltered[filter,]
+  
+  # filter by basin
+  if(!is.null((basin))){
+    filter <- which(dataBaseFiltered[,fields["basin"]] == basin)
+    dataBaseFiltered <- dataBaseFiltered[filter,]
+  }
+  
+  
+  # Get dimensions
+  sid <- dataBaseFiltered[,fields["sid"]]
+  countObs <- table(sid) # count of observation by storm
+  row <- max(countObs) # maximum number of rows
+  len <- length(unique(sid)) # number of unique storm in csv
+  cumulativeIndex <- cumsum(countObs) # starting points for each storm in csv
+
+  # Initialize template structure
+  templateArray = array(NaN, dim=c(row,len))
+  
+  # Mandatory fields
+  names <- array(NaN, dim=len)
+  seasons <- array(NaN, dim=len)
+  isotimes <- templateArray
+  longitude <- templateArray
+  latitude <- templateArray
+  msw <- templateArray
+  
+  # Optional fields
+  sshs <- templateArray # TODO Remove later
+  rmw <- templateArray
+  pressure <- templateArray
+  poci <- templateArray
+  
+  for(i in 1:len){
+    
+    # Get rows observations boundaries
+    if(i!=1){
+      start = cumulativeIndex[i-1] + 1
+    }else{
+      start = 1
+    }
+    end = cumulativeIndex[i]
+    
+    # Fill data
+    
+    names[i] <- dataBaseFiltered[start, fields["names"]]
+    seasons[i] <- as.numeric(dataBaseFiltered[start, fields["seasons"]])
+    isotimes[,i] <- c(dataBaseFiltered[start:end, fields["isoTime"]], rep(NaN, row-countObs[i]))
+    longitude[,i] <- as.numeric(c(dataBaseFiltered[start:end, fields["lon"]], rep(NaN, row-countObs[i])))
+    latitude[,i] <- as.numeric(c(dataBaseFiltered[start:end, fields["lat"]], rep(NaN, row-countObs[i])))
+    
+    if (unitConversion["msw"] == "mph2ms") {
```

Comment:
To avoid redundancy, I would first read all data either from netcdf or csv file, and then write a common 'unitConvert' function that takes the data and 'unitConversion' as input and gives the data converted as output

Reply:
Agree, It could be done, but I will focus on the others issues/tasks first